lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,278 @@
+    #
+    # Inputs: 1. ip:               IP address to be checked
+    #         2. subnet:           The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24",
+    #                              IPv6 - "2001:db8:abcd:12::/124")
+    #         3. group(optional):  The name of the External IPAM group containing the subnet to pull IP from
+    #
+    # Returns: true if IP exists in External IPAM, otherwise false.
+    #
+    # Responses from Proxy plugin:
+    #   Response if IP is already reserved:
+    #     {"code": 200, "exists": true}
+    #   Response if IP address is available
+    #     {"code": 404, "exists": false}
+    #   Response if missing required parameters:
+    #     {"code": 400, "error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)", "Missing 'ip' parameter. An IPv4 address must be provided(e.g. 100.10.10.22)"]}
+    #   Response if subnet not exists:
+    #     {"code": 404, "error": "No subnets found"}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def ip_exists(ip, subnet, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse get("/subnet/#{subnet}/#{ip}?group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['code'] == 200
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain IP address for subnet_id %s from External IPAM."), subnet)
+    end
+
+    # Deletes an IP address from a given subnet in External IPAM.
+    #
+    # Inputs: 1. ip:               IP address to be freed up
+    #         2. subnet:           The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24",
+    #                              IPv6 - "2001:db8:abcd:12::/124")
+    #         3. group(optional):  The name of the External IPAM group containing the subnet.
+    #
+    # Returns: true if IP is deleted successfully from External IPAM, otherwise false.
+    #
+    # Proxy responses:
+    #   Response if success:
+    #     {"code": 200}
+    #   Response if subnet error:
+    #     {"code": 404, "error": "The specified subnet does not exist in External IPAM."}
+    #   Response if IP already deleted:
+    #     {"code": 404, "error": "No addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def delete_ip_from_subnet(ip, subnet, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse delete("/subnet/#{subnet}/#{ip}?group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['code'] == 200
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to delete IP %s from the subnet %s in External IPAM."), ip, subnet)
```

Comment:
Beware our i18n code (gettext) is unable to handle two `%s` replacement wildcards. You have to do this:

    raise ProxyException.new(url, e, N_("Unable to delete IP %{ip} from the subnet %{subnet} in External IPAM."), ip: ip, subnet: subnet)

Please check the rest of this file, and possibly if you use ProxyException or WrappedException or Foreman::Exception they are all same.